crypto/rsa/rsa_asn1.c

```diff
@@ -27,10 +27,40 @@ static int rsa_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
         RSA_free((RSA *)*pval);
         *pval = NULL;
         return 2;
+    } else if (operation == ASN1_OP_D2I_POST) {
+        if (((RSA *)*pval)->version != RSA_ASN1_VERSION_MULTI) {
+            /* not a multi-prime key, skip */
+            return 1;
+        }
+        return (rsa_multip_calc_product((RSA *)*pval) == 1) ? 2 : 0;
     }
     return 1;
 }
 
+/* use customized new and free operations */
+static int rsa_prime_info_cb(int operation, ASN1_VALUE **pval,
+                             const ASN1_ITEM *it, void *exarg)
+{
+    if (operation == ASN1_OP_NEW_PRE) {
+        *pval = (ASN1_VALUE *)rsa_multip_info_new();
+        if (*pval != NULL)
+            return 2;
+        return 0;
+    } else if (operation == ASN1_OP_FREE_PRE) {
+        rsa_multip_info_free((RSA_PRIME_INFO *)*pval);
+        *pval = NULL;
+        return 2;
+    }
+    return 1;
+}
+
+/* Based on definitions in RFC 8017 appendix A.1.2 */
+ASN1_SEQUENCE_cb(RSA_PRIME_INFO, rsa_prime_info_cb) = {
+        ASN1_SIMPLE(RSA_PRIME_INFO, r, CBIGNUM),
+        ASN1_SIMPLE(RSA_PRIME_INFO, d, CBIGNUM),
+        ASN1_SIMPLE(RSA_PRIME_INFO, t, CBIGNUM),
+} ASN1_SEQUENCE_END_cb(RSA_PRIME_INFO, RSA_PRIME_INFO)
+
```

Comment:
None of this low level stuff in rsa_prime_info_cb should be necessary. You should just be able to make the new fields optional and extend RSA_free() to free up any multiprimes present.

Reply:
I don't quite understand this. It seems we also need to handle the `new` side things if there is no `rsa_prime_info_cb`; otherwise, who allocates those multi-prime related stuffs' memories in the RSA structure?

Reply:
The normal ASN.1 functions provide their own new/free methods based on the structure. The defaults may be sufficient in this case. If so that would mean the callback is not necessary. There may be something I'm missing though; I'll check.

Reply:
Ah... there is a _pp_ field which isn't part of the ASN.1 which is calculated and handled separately. While it is possible to allocate _pp_ as part of rsa_multip_calc_product() and delete rsa_prime_info_cb I don't see any advantage over doing this in the callback.

Reply:
One advantage is `rsa_multip_calc_product()` can do calculation only and allocation is done outside this function by the caller. This will make the code more clear to understand. But anyway, this time I put an allocation of `pp` in `rsa_multip_calc_product()` and removed the callback in ASN.1.